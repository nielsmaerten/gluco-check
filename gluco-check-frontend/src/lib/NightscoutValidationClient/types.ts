export type NightscoutValidationClientOptions = {
  endpointUrl: string;
};
