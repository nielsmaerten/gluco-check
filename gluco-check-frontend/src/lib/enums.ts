export enum BloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol/L",
}

export enum NightscoutBloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol",
}

// @TODO: once this repo is properly part of the workspace,
// pull this in from gluco-check-core instead
// These are mapped to translations directly by case name,
// additions or changes here need to be reflected there
export enum DiabetesMetric {
  Everything = "everything",
  BloodSugar = "blood sugar",
  InsulinOnBoard = "insulin on board",
  CarbsOnBoard = "carbs on board",
  SensorAge = "sensor age",
  CannulaAge = "cannula age",
  PumpBattery = "pump battery",
  PumpReservoir = "pump reservoir",
}

export enum AvailableLanguage {
  English = "en",
  Dutch = "nl",
}
