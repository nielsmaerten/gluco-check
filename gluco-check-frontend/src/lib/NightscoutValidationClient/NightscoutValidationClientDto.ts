import {
  IsBoolean,
  IsString,
  IsEnum,
  ValidateNested,
  ValidateIf,
} from "class-validator";
import { Type } from "class-transformer";
import { DiabetesMetric, NightscoutBloodGlucoseUnit } from "../enums";

export const returnNightscoutValidationEndpointResponseUrlType = () =>
  NightscoutValidationEndpointResponseUrl;
export const returnNightscoutValidationEndpointResponseTokenType = () =>
  NightscoutValidationEndpointResponseToken;
export const returnNightscoutValidationEndpointResponseNightscoutType = () =>
  NightscoutValidationEndpointResponseNightscout;

export const glucoseUnitValidationCondition = (o: any) => o.glucoseUnit !== "";

export class NightscoutValidationEndpointResponseUrl {
  @IsBoolean()
  isValid!: boolean;

  @IsString()
  parsed!: string;

  @IsBoolean()
  pointsToNightscout!: boolean;
}

export class NightscoutValidationEndpointResponseToken {
  @IsBoolean()
  isValid!: boolean;

  @IsString()
  parsed!: string;
}

export class NightscoutValidationEndpointResponseNightscout {
  @ValidateIf(glucoseUnitValidationCondition)
  @IsEnum(NightscoutBloodGlucoseUnit)
  glucoseUnit!: NightscoutBloodGlucoseUnit;

  @IsString()
  minSupportedVersion!: string;

  @IsString()
  version!: string;
}

export class NightscoutValidationEndpointResponseDto {
  @Type(returnNightscoutValidationEndpointResponseUrlType)
  @ValidateNested()
  url!: NightscoutValidationEndpointResponseUrl;

  @Type(returnNightscoutValidationEndpointResponseTokenType)
  @ValidateNested()
  token!: NightscoutValidationEndpointResponseToken;

  @Type(returnNightscoutValidationEndpointResponseNightscoutType)
  @ValidateNested()
  nightscout!: NightscoutValidationEndpointResponseNightscout;

  @IsEnum(DiabetesMetric, { each: true })
  discoveredMetrics!: DiabetesMetric[];
}
