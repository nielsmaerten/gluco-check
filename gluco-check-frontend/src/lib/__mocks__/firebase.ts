import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { GlucoCheckUserDocument } from "../types";

export const mockUser: Partial<firebase.User> = {
  email: "example@example.com",
  uid: "uid123",
};

export const mockUserDocument: GlucoCheckUserDocument = {
  heardDisclaimer: false,
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    url: "https://example.com",
    token: "nstoken123",
  },
};
