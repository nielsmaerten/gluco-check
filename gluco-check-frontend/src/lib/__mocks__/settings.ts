import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { SettingsFormData } from "../types";

export const mockFormData: SettingsFormData = {
  defaultMetrics: [DiabetesMetric.BloodSugar],
  nightscoutToken: "I am a token",
  nightscoutUrl: "https://example.com",
  glucoseUnit: BloodGlucoseUnit.mgdl,
};
