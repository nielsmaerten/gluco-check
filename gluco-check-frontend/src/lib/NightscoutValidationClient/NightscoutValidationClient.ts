import { validate } from "class-validator";
import { plainToClass } from "class-transformer";
import { NightscoutValidationClientOptions } from "./types";
import { NightscoutValidationEndpointResponseDto } from "./NightscoutValidationClientDto";
import { NightscoutValidationEndpointRequest } from "../types";

export const NSV_PATH = "/api/v1/validate-nightscout";
export const STATUS_OK = 200;

export class NightscoutValidationClient {
  private endpointUrl: string;
  private controller: AbortController | null;
  private signal: AbortSignal | null;
  private requestInProgress: boolean;

  constructor(options: NightscoutValidationClientOptions) {
    this.endpointUrl = options.endpointUrl;
    this.controller = new AbortController();
    this.signal = this.controller.signal;
    this.requestInProgress = false;
  }

  async fetchValidationStatus(
    nightscoutUrl: string,
    nightscoutToken: string
  ): Promise<NightscoutValidationEndpointResponseDto> {
    if (this.requestInProgress === true) {
      this.cancel();
    }
    try {
      this.requestInProgress = true;
      const requestBody: NightscoutValidationEndpointRequest = {
        url: nightscoutUrl,
        token: nightscoutToken,
      };
      const requestBodyRaw = JSON.stringify(requestBody);
      const headers = new Headers();
      headers.append("Content-Type", "application/json; charset=utf-8");

      let response;
      try {
        response = await fetch(`${this.endpointUrl}${NSV_PATH}`, {
          method: "POST",
          body: requestBodyRaw,
          redirect: "follow",
          headers,
          signal: this.signal,
        });
        if (response.status !== 200) {
          throw new Error("Received non-200 response status");
        }
      } catch (e) {
        throw new Error(`Failed to fetch from nsv endpoint: ${e}`);
      }

      const responseJson = await response?.json();
      try {
        const classifiedResponse = plainToClass(
          NightscoutValidationEndpointResponseDto,
          responseJson
        );

        const errors = await validate(classifiedResponse);
        if (errors.length) {
          throw errors;
        }

        return classifiedResponse;
      } catch (e) {
        throw new Error(`Invalid response from nsv endpoint: ${e}`);
      }
    } catch (e) {
      throw e;
    } finally {
      this.requestInProgress = false;
    }
  }

  async cancel() {
    this.controller?.abort();

    this.controller = null;
    this.signal = null;

    this.controller = new AbortController();
    this.signal = this.controller.signal;
  }

  requestIsInProgress(): boolean {
    return this.requestInProgress;
  }
}
