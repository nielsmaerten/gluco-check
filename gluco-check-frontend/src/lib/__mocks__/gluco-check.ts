import { plainToClass } from "class-transformer";
import { NightscoutValidationEndpointResponseDto } from "../NightscoutValidationClient/NightscoutValidationClientDto";

export const MOCK_NSV_RESPONSE_VALID = {
  token: {
    isValid: true,
    parsed: "token123",
  },
  url: {
    isValid: true,
    parsed: "https://example.com",
    pointsToNightscout: true,
  },
  nightscout: {
    glucoseUnit: "mmol",
    minSupportedVersion: "13.0.0",
    version: "14.0.7",
  },
  discoveredMetrics: [
    "blood sugar",
    "insulin on board",
    "carbs on board",
    "sensor age",
    "cannula age",
    "pump battery",
  ],
};

export const MOCK_NSV_RESPONSE_NON_NS_URL = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: false,
  },
  token: {
    parsed: "",
    isValid: false,
  },
  nightscout: {
    version: "",
    glucoseUnit: "",
    minSupportedVersion: "13.0.0",
  },
  discoveredMetrics: [],
};

export const MOCK_NSV_RESPONSE_NS_NEEDS_UPGRADE = {
  token: {
    isValid: false,
    parsed: "token123",
  },
  url: {
    isValid: true,
    parsed: "https://example.com",
    pointsToNightscout: true,
  },
  nightscout: {
    glucoseUnit: "mmol",
    minSupportedVersion: "13.0.0",
    version: "12.9.9",
  },
  discoveredMetrics: [],
};

export const MOCK_NSV_RESPONSE_INVALID_TOKEN = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: true,
  },
  token: {
    isValid: false,
    parsed: "token1234",
  },
  nightscout: {
    minSupportedVersion: "13.0.0",
    glucoseUnit: "mmol",
    version: "14.0.7",
  },
  discoveredMetrics: ["blood sugar"],
};

export const mockNsvResponseDtoValid = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_VALID
);

export const mockNsvResponseDtoNonNsUrl = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_NON_NS_URL
);

export const mockNsvResponseDtoNsNeedsUpgrade = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_NS_NEEDS_UPGRADE
);

export const mockNsvResponseDtoInvalidToken = plainToClass(
  NightscoutValidationEndpointResponseDto,
  MOCK_NSV_RESPONSE_INVALID_TOKEN
);
