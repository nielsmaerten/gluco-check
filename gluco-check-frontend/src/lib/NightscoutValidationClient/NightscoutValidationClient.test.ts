import { enableFetchMocks } from "jest-fetch-mock";
import { clone } from "ramda";
import { NightscoutBloodGlucoseUnit } from "../enums";
import { MOCK_NSV_RESPONSE_VALID } from "../__mocks__/gluco-check";
import { NightscoutValidationClient } from "./NightscoutValidationClient";

enableFetchMocks();

const MOCK_ENDPOINT_URL = "http://example.com";
const MOCK_NS_URL = "http://example.com/ns";
const MOCK_NS_TOKEN = "token123";

describe("NightscoutValidationClient", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it("Fetches validation status from endpoint", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(MOCK_NSV_RESPONSE_VALID));

    const nsvClient = new NightscoutValidationClient({
      endpointUrl: MOCK_ENDPOINT_URL,
    });
    const response = await nsvClient.fetchValidationStatus(
      MOCK_NS_URL,
      MOCK_NS_TOKEN
    );

    expect(fetchMock.mock.calls.length).toEqual(1);
    expect(fetchMock.mock.calls[0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });

  it("Fetches validation status from endpoint when units are mgdl", async () => {
    const responseWithMgdl = clone(MOCK_NSV_RESPONSE_VALID);
    responseWithMgdl.nightscout.glucoseUnit = NightscoutBloodGlucoseUnit.mgdl;
    fetchMock.mockResponseOnce(JSON.stringify(responseWithMgdl));

    const nsvClient = new NightscoutValidationClient({
      endpointUrl: MOCK_ENDPOINT_URL,
    });
    const response = await nsvClient.fetchValidationStatus(
      MOCK_NS_URL,
      MOCK_NS_TOKEN
    );

    expect(fetchMock.mock.calls.length).toEqual(1);
    expect(fetchMock.mock.calls[0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  });

  it("Throws if fetch throws", async () => {
    expect.assertions(1);
    fetchMock.mockRejectOnce(new Error("The lesson is never try."));

    const nsvClient = new NightscoutValidationClient({
      endpointUrl: MOCK_ENDPOINT_URL,
    });

    await expect(
      nsvClient.fetchValidationStatus(MOCK_NS_URL, MOCK_NS_TOKEN)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Failed to fetch from nsv endpoint: Error: The lesson is never try."`
    );
  });

  it("Throws if non-ok response returned", async () => {
    const responseWithMgdl = clone(MOCK_NSV_RESPONSE_VALID);
    responseWithMgdl.nightscout.glucoseUnit = NightscoutBloodGlucoseUnit.mgdl;
    fetchMock.mockResponseOnce(JSON.stringify(responseWithMgdl), {
      status: 500,
    });

    const nsvClient = new NightscoutValidationClient({
      endpointUrl: MOCK_ENDPOINT_URL,
    });

    await expect(
      nsvClient.fetchValidationStatus(MOCK_NS_URL, MOCK_NS_TOKEN)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Failed to fetch from nsv endpoint: Error: Received non-200 response status"`
    );
  });

  it("Throws if response validation fails", async () => {
    const invalidResponse = clone(MOCK_NSV_RESPONSE_VALID);
    invalidResponse.nightscout.glucoseUnit = "NOTAUNIT";
    fetchMock.mockResponseOnce(JSON.stringify(invalidResponse));

    const nsvClient = new NightscoutValidationClient({
      endpointUrl: MOCK_ENDPOINT_URL,
    });

    await expect(nsvClient.fetchValidationStatus(MOCK_NS_URL, MOCK_NS_TOKEN))
      .rejects.toThrowErrorMatchingInlineSnapshot(`
"Invalid response from nsv endpoint: An instance of NightscoutValidationEndpointResponseDto has failed the validation:
 - property nightscout.glucoseUnit has failed the following constraints: isEnum 
"
`);
  });

  describe("cancel", () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });
    afterEach(() => {
      jest.useRealTimers();
    });
    it("Cancels request when cancel is called", async () => {
      expect.assertions(1);
      jest.useFakeTimers();
      fetchMock.mockResponse(async () => {
        jest.advanceTimersByTime(10);
        return JSON.stringify(MOCK_NSV_RESPONSE_VALID);
      });

      const nsvClient = new NightscoutValidationClient({
        endpointUrl: MOCK_ENDPOINT_URL,
      });
      setTimeout(() => nsvClient.cancel(), 5);
      await expect(
        nsvClient.fetchValidationStatus(MOCK_NS_URL, MOCK_NS_TOKEN)
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"Failed to fetch from nsv endpoint: AbortError: The operation was aborted. "`
      );
      jest.useRealTimers();
    });
  });
});
