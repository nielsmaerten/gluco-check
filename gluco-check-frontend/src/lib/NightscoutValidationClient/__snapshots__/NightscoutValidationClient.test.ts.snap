// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NightscoutValidationClient Fetches validation status from endpoint 1`] = `
Array [
  "http://example.com/api/v1/validate-nightscout",
  Object {
    "body": "{\\"url\\":\\"http://example.com/ns\\",\\"token\\":\\"token123\\"}",
    "headers": Headers {
      Symbol(map): Object {
        "Content-Type": Array [
          "application/json; charset=utf-8",
        ],
      },
    },
    "method": "POST",
    "redirect": "follow",
    "signal": AbortSignal {},
  },
]
`;

exports[`NightscoutValidationClient Fetches validation status from endpoint 2`] = `
NightscoutValidationEndpointResponseDto {
  "discoveredMetrics": Array [
    "blood sugar",
    "insulin on board",
    "carbs on board",
    "sensor age",
    "cannula age",
    "pump battery",
  ],
  "nightscout": NightscoutValidationEndpointResponseNightscout {
    "glucoseUnit": "mmol",
    "minSupportedVersion": "13.0.0",
    "version": "14.0.7",
  },
  "token": NightscoutValidationEndpointResponseToken {
    "isValid": true,
    "parsed": "token123",
  },
  "url": NightscoutValidationEndpointResponseUrl {
    "isValid": true,
    "parsed": "https://example.com",
    "pointsToNightscout": true,
  },
}
`;

exports[`NightscoutValidationClient Fetches validation status from endpoint when units are mgdl 1`] = `
Array [
  "http://example.com/api/v1/validate-nightscout",
  Object {
    "body": "{\\"url\\":\\"http://example.com/ns\\",\\"token\\":\\"token123\\"}",
    "headers": Headers {
      Symbol(map): Object {
        "Content-Type": Array [
          "application/json; charset=utf-8",
        ],
      },
    },
    "method": "POST",
    "redirect": "follow",
    "signal": AbortSignal {},
  },
]
`;

exports[`NightscoutValidationClient Fetches validation status from endpoint when units are mgdl 2`] = `
NightscoutValidationEndpointResponseDto {
  "discoveredMetrics": Array [
    "blood sugar",
    "insulin on board",
    "carbs on board",
    "sensor age",
    "cannula age",
    "pump battery",
  ],
  "nightscout": NightscoutValidationEndpointResponseNightscout {
    "glucoseUnit": "mg/dl",
    "minSupportedVersion": "13.0.0",
    "version": "14.0.7",
  },
  "token": NightscoutValidationEndpointResponseToken {
    "isValid": true,
    "parsed": "token123",
  },
  "url": NightscoutValidationEndpointResponseUrl {
    "isValid": true,
    "parsed": "https://example.com",
    "pointsToNightscout": true,
  },
}
`;
